libethereum/EthereumHost.cpp

```diff
@@ -512,26 +513,31 @@ tuple<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>, vector
     vector<shared_ptr<EthereumPeer>> chosen;
     vector<shared_ptr<EthereumPeer>> allowed;
     vector<shared_ptr<SessionFace>> sessions;
-
-    size_t peerCount = 0;
+    double percentDecimal = _percent / 100.0;
     foreachPeer([&](std::shared_ptr<EthereumPeer> _p)
     {
         if (_allow(_p.get()))
         {
-            allowed.push_back(_p);
             sessions.push_back(_p->session());
+            allowed.push_back(_p);
         }
-        ++peerCount;
         return true;
     });
 
-    size_t chosenSize = (peerCount * _percent + 99) / 100;
+    if (_percent == 0 || allowed.size() == 0)
+    {
+        return make_tuple(move(chosen), move(allowed), move(sessions));
+    }
+
+    std::shuffle(allowed.begin(), allowed.end(), m_urng);
+
+   // Remove elements from the end of the shuffled allowed vector and move them to chosen.
+    size_t chosenSize = percentDecimal * allowed.size();
     chosen.reserve(chosenSize);
-    for (unsigned i = chosenSize; i && allowed.size(); i--)
+    for (auto it = allowed.end(); it >= allowed.begin() + chosenSize; it--)
```

Comment:
Still I think you don't need a loop for this
```suggestion
    std::move(allowed.begin() + chosenSize, allowed.end(), std::back_inserter(chosen)); 
    allowed.erase(allowed.begin() + chosenSize, allowed.end());
```